### PR TITLE
Lowmem dropout latest - DRAFT - DO NOT MERGE

### LINKED ISSUE #102319: Re-enable Low Memory Dropout
### 🐛 Describe the bug

https://github.com/pytorch/pytorch/pull/100064 caused 3% perf dropout in HF training because of disabled low memory dropout

### Versions

master

### PR DESCRIPTION
**Pull request by amritahs-ibm**
Fixes https://github.com/pytorch/pytorch/issues/102319 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

### CODE REVIEW

**Comment by amritahs-ibm** [comment]
PFA the kernels generated during forward and backward pass on GPU for testcases test_dropout2, test_dropout3 and test_dropout4. 

This is the summary:
image


[test_dropout2_gpu_base_kernels.txt](https://github.com/user-attachments/files/23933461/test_dropout2_gpu_base_kernels.txt)
[test_dropout2_gpu_patch_kernels.txt](https://github.com/user-attachments/files/23933463/test_dropout2_gpu_patch_kernels.txt)
[test_dropout3_gpu_base_kernels.txt](https://github.com/user-attachments/files/23933465/test_dropout3_gpu_base_kernels.txt)
[test_dropout3_gpu_patch_kernels.txt](https://github.com/user-attachments/files/23933467/test_dropout3_gpu_patch_kernels.txt)
[test_dropout4_gpu_base_kernels.txt](https://github.com/user-attachments/files/23933468/test_dropout4_gpu_base_kernels.txt)
[test_dropout4_gpu_patch_kernels.txt](https://github.com/user-attachments/files/23933470/test_dropout4_gpu_patch_kernels.txt)

File: torch/_functorch/partitioners.py [hunk 1]
```diff
@@ -1835,6 +1835,25 @@ def is_fusible(a, b):
             "Need networkx installed to perform smart recomputation heuristics"
         ) from e
 
+    def _is_regeneratable_from_seed(node):
```
Comment (eellison) [inline torch/_functorch/partitioners.py:1965]: why do we need this?

File: torch/_functorch/config.py [hunk 2]
```diff
@@ -193,7 +195,7 @@ def remote_autograd_cache_default() -> Optional[bool]:
 # Sets all of the ban_recompute heuristics to False except ban_recompute_reductions
 # Generally, this will probably result in some memory improvement, but at the
 # cost of some performance
-aggressive_recomputation = False
+aggressive_recomputation = partitioner_aggressive_fusion
```
Comment (eellison) [inline torch/_functorch/config.py:221]: please limit unrelated changes

File: torch/_functorch/partitioners.py [hunk 3]
```diff
@@ -1860,6 +1879,8 @@ def should_ban_recomputation(node):
             return False
         if node.target in [aten.lift_fresh_copy.default, aten.lift_fresh.default]:
             return False
+        if inductor_config.lowmem_dropout and _is_regeneratable_from_seed(node):
```
Comment (eellison) [inline torch/_functorch/partitioners.py:2010]: Can you say more about this

Comment (eellison) [reply torch/_functorch/partitioners.py:2010]: If any node is not recomputable, then nodes dependent on it should not be recomputed either.

**Review by eellison — COMMENTED** [review]
looks reasonable - just submitting perf dashboard run

File: torch/_functorch/partitioners.py (hunk 1 above)
Comment (amritahs-ibm) [reply torch/_functorch/partitioners.py:1965]: This method checks if mask can be computed deterministically from a given seed.
So if a node is inductor_random or derived from inductor_random node, then we will return True, indicating that mask can be regenerated from seed.

File: torch/_functorch/partitioners.py (hunk 3 above)
Comment (amritahs-ibm) [reply torch/_functorch/partitioners.py:2010]: If lowmem_dropout flag is set and if we are sure of regenerating mask from the seed, then we are enabling recomputation of mask. This ensures that mask is not saved on first call to inductor_random, instead mask will be recomputed everytime.
